airflow/contrib/hooks/gcp_dataflow_hook.py

```diff
@@ -124,36 +127,38 @@ def __init__(self, cmd):
 
     def _line(self, fd):
         if fd == self._proc.stderr.fileno():
-            lines = self._proc.stderr.readlines()
-            for line in lines:
-                self.log.warning(line[:-1])
-            if lines:
-                return lines[-1]
+            return self._proc.stderr.readline()
         if fd == self._proc.stdout.fileno():
-            line = self._proc.stdout.readline()
-            return line
+            return self._proc.stdout.readline()
 
     @staticmethod
     def _extract_job(line):
-        if line is not None:
-            if line.startswith("Submitted job: "):
-                return line[15:-1]
+        job_id_pattern = re.compile(
+            '.*https://console.cloud.google.com/dataflow.*/jobs/([a-z|0-9|A-Z|\-|\_]+).*')
+        matched_job = job_id_pattern.match(line or '')
+        if matched_job:
+            return matched_job.group(1)
 
     def wait_for_done(self):
         reads = [self._proc.stderr.fileno(), self._proc.stdout.fileno()]
         self.log.info("Start waiting for DataFlow process to complete.")
-        while self._proc.poll() is None:
+        job_id = None
+        while True:
             ret = select.select(reads, [], [], 5)
             if ret is not None:
                 for fd in ret[0]:
                     line = self._line(fd)
                     if line:
-                        self.log.debug(line[:-1])
+                        self.log.info(line[:-1])
+                        job_id = job_id or self._extract_job(line)
             else:
                 self.log.info("Waiting for DataFlow process to complete.")
+            if self._proc.poll() is not None:
```

Comment:
Is this here now to prevent final logs from getting dropped? A comment would help prevent this from being undone.

Reply:
Not really, it makes sure the job extraction logic is called when the subprocess terminates before coming to wait_for_done. self._proc.poll() is not None --> subprocess ends

Reply:
I think this may not always ensure that each STDERR/STDOUT line is processed.

For example, say a process completes ~instantly and logs 100 lines to STDERR and STDOUT.  The code as it is now seems like it would only process 1 line of each stream, then terminate. You'd need to read until the streams are empty (e.g. readlines). We should probably also ensure the streams are empty after the process terminates.

Reply:
Reading each of stdout and stderr independently is hard to do without deadlocking one or the other too -you can't just read one to the end then read the other. https://stackoverflow.com/questions/33886406/how-to-avoid-the-deadlock-in-a-subprocess-without-using-communicate

(Sorry if this is out of context - I haven't looked at the PR, just seen the comments)

Reply:
Thanks for the suggestion, note that we are not trying to read/write to STDIN/STDOUT. 
Revised to do one more round of stderr/stdout reading even after subprocess ends.